ViewEnvironmentUI/Sources/ViewEnvironmentPropagatingObject.swift

```diff
@@ -0,0 +1,272 @@
+        let object = NSObject()
+        needsUpdateObservers[object] = onNeedsUpdate
+        return .init { [weak self] in
+            self?.needsUpdateObservers[object] = nil
+        }
+    }
+
+    /// The [`ViewEnvironment` propagation](x-source-tag://ViewEnvironmentObserving) ancestor.
+    ///
+    /// This describes the ancestor that the `ViewEnvironment` is inherited from.
+    ///
+    /// To override the return value of this property, set the ``environmentAncestorOverride``.
+    /// If no override is present, the return value will be `defaultEnvironmentAncestor`.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public var environmentAncestor: ViewEnvironmentPropagating? {
+        environmentAncestorOverride?() ?? defaultEnvironmentAncestor
+    }
+
+    /// The [`ViewEnvironment` propagation](x-source-tag://ViewEnvironmentObserving)
+    /// descendants.
+    ///
+    /// This describes the descendants that will be notified when the `ViewEnvironment` changes.
+    ///
+    /// To override the return value of this property, set the ``environmentDescendantsOverride``.
+    /// If no override is present, the return value will be `defaultEnvironmentDescendants`.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public var environmentDescendants: [ViewEnvironmentPropagating] {
+        environmentDescendantsOverride?() ?? defaultEnvironmentDescendants
+    }
+
+    /// ## SeeAlso ##
+    /// - ``environmentAncestorOverride``
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public typealias EnvironmentAncestorProvider = () -> ViewEnvironmentPropagating?
+
+    /// This property allows you to override the propagation path of the `ViewEnvironment` as it flows through the
+    /// node hierarchy by overriding the return value of `environmentAncestor`.
+    ///
+    /// The result of this closure should be the propagation node that the `ViewEnvironment` is inherited from.
+    ///
+    /// If this value is `nil` (the default), the resolved value for the ancestor will be `defaultEnvironmentAncestor`.
+    ///
+    /// ## Important ##
+    ///  - You must not set overrides while overrides are already set—doing so will throw an assertion. This assertion
+    ///    prevents accidentally clobbering an existing propagation path customization defined somewhere out of your
+    ///    control (e.g. Modals customization).
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public var environmentAncestorOverride: EnvironmentAncestorProvider? {
+        get {
+            objc_getAssociatedObject(self, &AssociatedKeys.ancestorOverride) as? EnvironmentAncestorProvider
+        }
+        set {
+            assert(
+                newValue == nil
+                    || environmentAncestorOverride == nil,
+                "Attempting to set environment ancestor override when one is already set."
+            )
+            objc_setAssociatedObject(self, &AssociatedKeys.ancestorOverride, newValue, .OBJC_ASSOCIATION_RETAIN)
```

Comment:
Trying to think through if we can get into any weird cases by making this a retain, vs a weak.

Reply:
The type here is a closure so it'd be really easy to mess that up I think...

The fact that this property is retained is part of the trick I used in the prototype to add support for inserting a propagation node above the view controller being described in a `ViewControllerDescription`—this closure is the only thing holding a reference to that fake node which provides the bridged `ViewEnvironment`.